grep-0001-coding-guidelines.md

```diff
@@ -0,0 +1,122 @@
+editor, and there is no reason to not adopt these settings. This makes it
+easier for multiple people to contribute to the same codebase, and will keep
+commits free of non-functional changes.
+
+To summarize, consistency is extremely useful, and coding guidelines help ensure
+consistency. We assume that all contributors are mature enough to at least try
+and follow them.
+
+## Description
+
+### General Coding Guidelines
+
+* Never submit code with trailing whitespace.
+* Code is read more often than it's written. Code readability is thus something
+  worth optimizing for.
+* Try and keep line lengths to 79 characters, unless readability suffers. We
+  often have to do fun things like SSH into machines and edit code in a
+  terminal, and do side-by-side views of code on small-ish screens, so this is
+  actually pretty helpful.
+
+### C++-specific Guidelines
+
+* If in doubt, consult the [C++ Core Guidelines][https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md].
+  If the guidelines have an answer, and it works for you, just pick that.
+* Use Doxygen doc-blocks copiously.
+* All things equal, prefer standard C++ constructs over Boost constructs (see
+  also Boost guidelines).
+* Given the option, prefer C++ lambdas over std::bind, and just don't use
+  boost::bind if you can.
+* Include include files in the following order: Local headers, other GNU Radio
+  headers, 3rd-party library headers, Boost headers, standard headers.
+  The rationale is to include from most to least specific. This is the best way
+  to catch missing includes (if you were to include the standard header first,
+  it would be available to all include files that come later. If they need that
+  standard header too, they should be including it themselves).
+  Example:
+
+```cpp
+#include "module_specific.hpp"
+#include <gnuradio/block.h>
+#include <fftw3.h>
+#include <boost/shared_ptr.hpp>
+#include <mutex>
+```
+
+* Feel free to use modern C/C++ features even if they were not used before.
+  Make sure they work with the compilers and dependencies which are set for the
+  version of UHD the commit will be made upon. The GNU Radio CI system will be
+  able to confirm this. (Note: C++11 features are available starting with
+  version 3.8).
+
+
+## Boost-specific Guidelines
+
+* Avoid Boost where possible.
+* Don't use Boost's sleep functions. Prefer `std::chrono` functions.
+
+## Python-specific Guidelines
+
+* Follow the suggestions in PEP8 (https://www.python.org/dev/peps/pep-0008/)
+  and PEP257 (https://www.python.org/dev/peps/pep-0257/). The former is about
+  code layout in general, the latter about docstrings.
+* Keep Python code compatible with Py2k and Py3k, where possible.
+* Pylint is good tool for helping with following code guidelines. It's very
+  fussy though, so don't get too worked up about following its suggestions.
+
+## Revision Control Hygiene
+
+* In this repository, we almost always use fast-forward merges, and no merge
```

Comment:
In my experience, when writing guidelines, it's better to give the reasoning behind, so readers know "ah, that's why they insist on that". See your terminal-usage example above, which is great.

What is the reasoning here (no merge commits)?

Reply:
So, this is one of the guidelines I just copied over from UHD. There's two things here:
- Fast-forward merges: These are best to keep development linear, and add new features always on top. Also, the best way to avoid merge conflicts (but that's already somewhat subjective). This one I feel pretty strongly about.
- No merge commits: When doing FF merges, they are simply not necessary and clutter the history. However, they can also be nice because they give some indication of the source branch. git default is to not do merge commits on ff merges, but we could do this differently.

Reply:
...but excellent point about adding rationales. I think we could use more of those.

Reply:
Regarding the 2nd issue, I agree with 2 of your statements: if doing FF merges, merge commits are less useful, but still nice. But, it is completely based on the 1st issue.

I understand the history cleanliness that comes with FF merges, but for them to work, it means the collaborator needs to rebase his branch (assuming the source branch, e.g., master, progressed since branching), which usually causes more conflicts, and forces her/him to modify the history - an unpleasant action of several collaborators work on the same branch, and in general, and in addition, you lose information.

The approach I find most useful:

* Always have a merge commit.
* Don't enforce rebasing. If a fast-forward is possible (master hasn't evolved), so be it. Otherwise, fast-forward isn't a must.
* Collaborator solves conflicts by merging master into her own branch when they finish (actually, preferred on an ongoing basis, as it makes solving conflicts much easier) and solve all of the conflicts. That way, when the project maintainer merges back into master, there are no conflicts.
* Commit comment guidelines (prefixes, components, etc.) aren't strictly enforced on commits that belong to a branch, but are encouraged for important commits. This allows the developer more freedom and flexibility, e.g., commit something because they are going to continue next week and don't want to risk losing it, so they just use a message of "temp", "comment" if it's a slight comment change, "docs", etc.
* Commit comment guidelines are enforced on merge commits on master.
* When developers want to see the history with its most important commits (AKA only the merge commits), they use `git log --first-parent`. They get nice, linear, well-documented history. When they want to see the small details and try to simulate the way of thinking of the developer, they can check the branch commits.

Less related to this issue:

* Merge commits should include the branch name and a link to GitHub's PR.
* When solving conflicts and committing, it's best to document in the commit message the files that required conflict resolution (I think you get it as a commented-out list anyway) - this help in tracking bugs due to bad conflict resolutions ("oh, he merged and solved a conflict, but didn't realize line A should come before B").

Reply:
I'd like to hear more people's opinions on this. Like I said, this was a base for discussion. A couple of things on your comments though:

- Your approach still doesn't aid bisecting
- When commits are submitted for merge, the commit message should be enforced (because they end up in the final history). People's intermediate commits need to be cleaned up before submitting a PR. And before submitting your PR, you should rebase anyway, that's just good manners. The developer doesn't necessarily have to do the final rebase, the maintainer can still do that (it's even built into github UI these days) and resolving merge conflicts is either easy (in which case the maintainer can do it) or not (in which case the original developer needs to do it anyway, regardless of it being a rebase or a regular merge).

Basically, you're saying all the hard work needs to be with the maintainer, and the developer can do pretty much what they want, including formatting commit messages -- that's not acceptable.

Reply:
Why does my approach require rebasing? I might be missing something.

Reply:
Your approach does not require rebasing, but that's what I'm complaining about 😄 Before a PR gets submitted, people should rebase, that's just good manners and is the best way to ensure no merge conflicts. Criss-cross merges (i.e., first merging master into your branch) means we end up with even more merge commits, and bisects become even harder.

I like some of your points, though. If we have merge commits, then references to the source branch are a good idea. Maybe this would be better served on the mailing list, where more people see this discussion and can chime in.

Reply:
Regarding rebases:

I was very tempted to phrase it quite a bit stricter than what @mbr0wn put into that document. Again, this could do very well with a rationale, put into a separate bullet point:

> * Before opening a request for merging your code, make sure to rebase your changes against the current state of the branch you forked off your own branch (typically, that's `master`, so that procedure would amount to `git fetch origin; git rebase origin/master`). <br> That way, you keep control of how your code integrates into the common code base! If everything goes well, rebasing requires absolutely no interaction on your side. If something goes wrong, fix it like you'd fix a merge conflict. It's the same effort as fixing a merge conflict, and avoids one of those later on, but puts you in command over the order of lines, and enables you (who's the most knowledgeable expert on your own code) to spot problems that through naive merging by a maintainer would have become a bug.

Reply:
I've added this paragraph in the latest version.